pennylane/measurements/shots.py

```diff
@@ -0,0 +1,176 @@
+
+    One should also note that specifying a single tuple of length 2 is considered two different
+    shot values, and *not* a tuple-pair representing shots and copies to avoid special behaviour
+    depending on the iterable type:
+
+    >>> shots = Shots((100, 2))
+    >>> shots.total_shots, shots.shot_vector
+    (102, (ShotCopies(shots=100, copies=1), ShotCopies(shots=2, copies=1)))
+    """
+
+    total_shots: int = None
+    """The total number of shots to be executed."""
+
+    shot_vector: Tuple[ShotCopies] = None
+    """The tuple of :class:`~ShotCopies` to be executed. Each element is of the form ``(shots, copies)``."""
+
+    _SHOT_ERROR = ValueError(
+        "Shots must be a single positive integer, a tuple pair of the form (shots, copies), or a sequence of these types."
+    )
+
+    _frozen = False
+
+    def __new__(cls, shots):
+        return shots if isinstance(shots, cls) else object.__new__(cls)
+
+    def __setattr__(self, name, value):
+        if self._frozen:
+            raise AttributeError(
+                "Shots is an immutable class. Consider creating a new instance if you need different shot values."
+            )
+        return super().__setattr__(name, value)
+
+    def __init__(self, shots):
+        if shots is None:
+            self.total_shots = None
+            self.shot_vector = ()
+        elif isinstance(shots, int):
+            if shots < 1:
+                raise self._SHOT_ERROR
+            self.total_shots = shots
+            self.shot_vector = (ShotCopies(shots, 1),)
+        elif isinstance(shots, Sequence):
+            if not all(valid_int(s) or valid_tuple(s) for s in shots):
+                raise self._SHOT_ERROR
+            self.__all_tuple_init__([s if isinstance(s, tuple) else (s, 1) for s in shots])
+        elif isinstance(shots, self.__class__):
+            return  # self already _is_ shots as defined by __new__
+        else:
+            raise self._SHOT_ERROR
+
+        self._frozen = True
+
+    def __all_tuple_init__(self, shots: Sequence[Tuple]):
+        res = []
+        total_shots = 0
+        current_shots, current_copies = shots[0]
+        for s in shots[1:]:
+            if s[0] == current_shots:
+                current_copies += s[1]
+            else:
+                res.append(ShotCopies(current_shots, current_copies))
+                total_shots += current_shots * current_copies
+                current_shots, current_copies = s
+        self.shot_vector = tuple(res + [ShotCopies(current_shots, current_copies)])
+        self.total_shots = total_shots + current_shots * current_copies
+
+    @property
+    def has_partitioned_shots(self):
+        """
+        Evaluates to True if this instance represents either multiple shot
+        quantities, or the same shot quantity repeated multiple times.
+
+        Returns:
+            bool: whether or not shots are partitioned
+        """
+        return len(self.shot_vector) > 1 or self.shot_vector[0].copies > 1
```

Comment:
This will raise an error if `shots=None` since `self.shot_vector` will be empty.

```suggestion
        return self.total_shots and (len(self.shot_vector) > 1 or self.shot_vector[0].copies > 1)
```